### PR TITLE
Fix horrible lag in Zombies mode

### LINKED ISSUE #32870: Zombies cause massive framerate issues as AI
## Description
Playing AI, when crew starts to get converted to zombies, my framerate gradually drops when looking at zombies.

Framerate seems to be dependent on how many zombies are on screen. So I think it is the zombie characters themselves that are causing the drops, and not the large amount of bullet casings and blood that usually comes with a zombie round. 

**Reproduction**
1. Be AI
2. Zombie round with a bunch of players
3. Lots of crew get converted
4. SS14 powered by microsoft powerpoint

**Screenshots**

https://github.com/user-attachments/assets/3f696201-4f1a-4a25-8120-0d94f66ce89d

FPS in the top-right of the screen.

**Additional context**
I've heard other people in OOC confirm that they have this as AI, but I do not know if it is universal.

My specs, if at all relevant. It's a performance/framerate thing after all:

- AMD Ryzen 7 5800X3D
- AMD Radeon RX 6950 XT
- 32gb of ram whatever

### PR DESCRIPTION
**Pull request by psykana**
## About the PR
This fixes massive FPS drop in vicinity of zombified players.
Closes #32870.

## Why / Balance
Many FPS = good

## Technical details
With some experimentation I figured out lag is caused specifically by Zombie component, not a "complete" Zombie antag.
I spawned ~30 Urists in a box, gave them the Zombie component, and once lag settled in for real, turned on IDE profiling. Turns out ~75% of CPU time is spent in `Content.Shared.Whitelist.EntityWhitelistSystem.IsValid`, which was called from `StatusIconSystem`.
Logical error here blows up the `EntityWhitelist.Registrations` List to 100k+ duplicate elements which completely chokes the processor in the following loop.
Zombies (and Revolutionaries) are the only prototypes that use component lookup for `factionIcon`, so it didn't blow up anywhere else.
Fix is self-explanatory.

## Media
image_2024-12-11_03-22-50
image_2024-12-11_03-25-07

## Requirements
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
None

**Changelog**
:cl:
- fix: Massive FPS drop in vicinity of zombified players

### CODE REVIEW

File: Content.Shared/Whitelist/EntityWhitelistSystem.cs
Comment (metalgearsloth) [inline Content.Shared/Whitelist/EntityWhitelistSystem.cs:59]: This is the same code?

**Comment by slarticodefast** [comment]
Similar to https://github.com/space-wizards/space-station-14/pull/33322
Not sure which approach is better

File: Content.Shared/Whitelist/EntityWhitelistSystem.cs
Comment (psykana) [reply Content.Shared/Whitelist/EntityWhitelistSystem.cs:59]: Line 57 here keeps piling up the same `regs` onto existing List with every function call.

Comment (deltanedas) [reply Content.Shared/Whitelist/EntityWhitelistSystem.cs:59]: better yet put the `var regs` line in the block as well

**Comment by metalgearsloth** [comment]
> Related to #33322 Not sure which approach is better

I don't really like either PR but in lieu of having a better refactor I'd prefer this one.

**Comment by psykana** [comment]
Can we get this escalated to P1?
Quoting @metalgearsloth on Discord:
> atm on master it's making the game unplayable on its own so anything is going to be better

I see that maintainers decided to refactor this system, which is great, but I think it's not unreasonable to deploy this hotfix now AND work on a refactor.

File: Content.Shared/Whitelist/EntityWhitelistSystem.cs
Comment (stalengd) [inline Content.Shared/Whitelist/EntityWhitelistSystem.cs:52]: I feel that this should make this a bit safer. Unfortunately, we can't efficiently track each array element value with this API.
```suggestion
            if (list.Registrations == null || list.Registrations.Count != list.Components.Length)
```

**Review by stalengd — CHANGES_REQUESTED** [review]

File: Content.Shared/Whitelist/EntityWhitelistSystem.cs
Comment (psykana) [inline Content.Shared/Whitelist/EntityWhitelistSystem.cs:50]: It was brought to my attention by @stalengd that my change breaks API.
He's proposing the following solution:
```suggestion
        if (list.Components != null)
        {
            var regs = StringsToRegs(list.Components);

            list.Registrations ??= new List<ComponentRegistration>();
            list.Registrations.Clear();
            list.Registrations.AddRange(regs);
        }
```

**Comment by Errant-4** [comment]
Can we change the target to Stable or will that drag in some early updates from Master?

**Comment by VasilisThePikachu** [comment]
Hi @psykana

Due to your fork being based of serbia im unable to push to your branch ([thanks github](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork))

Please do the following (assuming upstream is wizden)
```
git fetch upstream
git rebase --onto upstream/stable upstream/master FIX-BRANCH
git push --force
```
Then switch your base branch on GitHub to stable (Click edit on the title -> Base: Stable)

**Comment by psykana** [comment]
Hi @VasilisThePikachu, should be fine now, sorry about that. Please fix labels.

**Comment by VasilisThePikachu** [comment]
> Hi @VasilisThePikachu, should be fine now, sorry about that. Please fix labels.

Not to worry. Thank you

**Comment by VasilisThePikachu** [comment]
I did just notice I broke maintainer policy because we needed 2 other approvals UHHH oopsy :3

**Pull request merged**
